python/src/pyarrow/common.cc

```diff
@@ -47,6 +47,20 @@ class PyArrowMemoryPool : public arrow::MemoryPool {
     return Status::OK();
   }
 
+  Status Reallocate(int64_t old_size, int64_t new_size, uint8_t** ptr) override {
+    *ptr = reinterpret_cast<uint8_t*>(std::realloc(*ptr, new_size));
+
+    if (*ptr == NULL) {
+      std::stringstream ss;
+      ss << "realloc of size " << new_size << " failed";
+      return Status::OutOfMemory(ss.str());
+    }
+
+    bytes_allocated_ += new_size - old_size;
+
+    return Status::OK();
+  }
```

Comment:
Perhaps we should simply compose the default Arrow memory pool (but accept any implementation) in this class and account for our own Python memory allocations (effectively this is a suballocator, then, but we haven't implemented general child allocators yet)? This way we can use whatever allocator we want in Python